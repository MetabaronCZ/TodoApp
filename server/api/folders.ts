import { Application } from 'express';
import { InferType, object, string } from 'yup';

import { Folder } from 'models/Folder';

import { mockedDb } from '../db';
import { mockApiRequest, mockCreatedId, mockParseParamArray } from '../utils';

const createFolderDataSchema = object({
  title: string().required(),
}).required();

const editFolderDataSchema = object({
  title: string(),
}).required();

export const setApiFolderEndpoints = (app: Application): void => {
  // get folder list
  app.get('/folder', async (req, res) => {
    await mockApiRequest();
    res.json({ data: mockedDb.folders });
  });

  // create folder
  app.post('/folder', async (req, res) => {
    const data = req.body;
    await mockApiRequest();

    try {
      await createFolderDataSchema.validate(data);
      const folderData = data as InferType<typeof createFolderDataSchema>;

      const folder: Folder = {
        ...folderData,
        id: mockCreatedId('FOLDER'),
      };
      mockedDb.folders = [...mockedDb.folders, folder];
      res.json({ data: folder });
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // edit folder
  app.patch('/folder/:id', async (req, res) => {
    const { id } = req.params;
    const data = req.body;
    await mockApiRequest();

    try {
      await editFolderDataSchema.validate(data);
      const folderData = data as InferType<typeof editFolderDataSchema>;

      mockedDb.folders = mockedDb.folders.map((item) => {
        return id === item.id ? { ...item, ...folderData } : item;
      });

      res.json(null);
    } catch (error) {
      res.status(500).json({ error });
    }
  });

  // delete folders
  app.delete('/folder', async (req, res) => {
    const { ids } = req.query;
    await mockApiRequest();

    const idsParsed = mockParseParamArray(ids);

    const validIds = mockedDb.folders
      .filter((item) => idsParsed.includes(item.id))
      .map((item) => item.id);

    if (0 === validIds.length) {
      res.status(500).json({ error: 'Could not delete folders: Invalid ID!' });
    } else {
      mockedDb.folders = mockedDb.folders.filter((item) => {
        return !validIds.includes(item.id);
      });

      mockedDb.todos = mockedDb.todos.filter((item) => {
        if (item.folder && validIds.includes(item.folder)) {
          return { ...item, folder: null };
        }
        return item;
      });

      res.json(null);
    }
  });
};
