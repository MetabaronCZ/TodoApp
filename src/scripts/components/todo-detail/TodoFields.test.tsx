import React from 'react';
import { Provider } from 'react-redux';

import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';

import * as Dropdown from 'components/forms/Dropdown';
import * as ItemList from 'components/common/ItemList';
import { TodoFields } from 'components/todo-detail/TodoFields';

import { TodoData } from 'models/Todo';
import { Folder } from 'models/Folder';
import { FormErrors } from 'hooks/useForm';
import { todoFoldersSlice } from 'store/folders';

import { mockStore } from 'test/store';
import { withMockedProviders } from 'test/component';

const { setFolders } = todoFoldersSlice.actions;

const testFoldersData: Folder[] = [
  { id: 'A', title: 'Folder A' },
  { id: 'B', title: 'Folder B' },
  { id: 'C', title: 'Folder C' },
];

const testData: TodoData = {
  title: 'Edited todo',
  description: '',
  isDone: true,
  folder: 'B',
};

interface TodoFieldsConfig {
  readonly loading?: boolean;
  readonly data?: TodoData;
  readonly errors?: FormErrors<TodoData>;
  readonly onChange?: () => void;
}

const getTodoFields = (config: TodoFieldsConfig = {}): JSX.Element => {
  const store = mockStore();
  store.dispatch(setFolders(testFoldersData));

  return withMockedProviders(
    <Provider store={store}>
      <TodoFields
        loading={config.loading}
        fields={config.data ?? testData}
        errors={config.errors}
        onChange={config.onChange ?? (() => null)}
      />
    </Provider>,
  );
};

describe('components/todo-detail/TodoFields', () => {
  it('should render correctly', () => {
    const { container } = render(getTodoFields());
    expect(container).toMatchSnapshot();
  });

  it('should render props', () => {
    const itemList = jest.spyOn(ItemList, 'ItemList');
    const dropdown = jest.spyOn(Dropdown, 'Dropdown');
    expect(itemList).toHaveBeenCalledTimes(0);
    expect(dropdown).toHaveBeenCalledTimes(0);

    const tree = render(
      getTodoFields({
        errors: { title: 'Title field error' },
        loading: true,
      }),
    );
    expect(tree.container).toMatchSnapshot();

    expect(itemList).toHaveBeenCalledTimes(1);
    expect(itemList.mock.calls[0][0].loading).toEqual(true);

    const inputs = tree.container.querySelectorAll('input');
    expect(inputs.length).toEqual(2);
    const [title, isDone] = Array.from(inputs);

    expect(title).toBeInTheDocument();
    expect(title?.value).toEqual('Edited todo');

    expect(isDone).toBeInTheDocument();
    expect(isDone?.checked).toEqual(true);

    expect(dropdown).toHaveBeenCalledTimes(1);
    expect(dropdown.mock.calls[0][0].value).toEqual('B');

    const description = tree.container.querySelector('textarea');
    expect(description).toBeInTheDocument();
    expect(description?.value).toEqual('');

    expect(tree.getByText('Title field error')).toBeInTheDocument();
  });

  it('should call onChange', () => {
    const onChange = jest.fn();

    const tree = render(getTodoFields({ onChange }));
    expect(onChange).toHaveBeenCalledTimes(0);

    const input = tree.container.querySelector('input');
    expect(input).toBeInTheDocument();

    // trigger onChange event
    if (input) {
      fireEvent.change(input, { target: { value: 'New title' } });
    }
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('title', 'New title');
  });
});
