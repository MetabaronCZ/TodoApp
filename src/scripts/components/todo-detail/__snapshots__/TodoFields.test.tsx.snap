// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/todo-detail/TodoFields > should render correctly 1`] = `
.c12 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c9 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c10 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c10:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c10:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c10:disabled {
  color: #c0c0c0;
}

.c11 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8 {
  position: relative;
  display: inline-block;
}

.c0 {
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  border-bottom: 1px solid #c0c0c0;
}

.c1:last-child {
  border-bottom: none;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: 30px;
}

.c14 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  width: 20px;
  height: 20px;
  border: 2px solid #222;
  color: #222;
  background: #fff;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.c14:hover {
  background-color: #fcfcfc;
}

.c14:focus-within {
  outline: 1px dotted #222;
}

.c15 {
  position: absolute;
  top: 0;
  left: 0;
  width: 0;
  height: 0;
  opacity: 0;
  outline: none;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c7 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
}

.c7:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c7:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c7:disabled {
  color: #c0c0c0;
}

.c17 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
  min-height: 110px;
  resize: vertical;
}

.c17:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c17:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c17:disabled {
  color: #c0c0c0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: block;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100px;
  min-height: 30px;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="form-field-0"
          >
            Název úkolu:
          </label>
        </div>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <input
              class="c7"
              id="form-field-0"
              maxlength="80"
              type="text"
              value="Edited todo"
            />
          </div>
        </div>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="form-field-1"
          >
            Složka:
          </label>
        </div>
        <div
          class="c5"
        >
          <div
            class="c8"
          >
            <button
              class="c9 c10"
              id="form-field-1"
              type="button"
            >
              <span
                class="c11"
              >
                Folder B
              </span>
              <i
                class="c12"
              >
                ˅
              </i>
            </button>
          </div>
        </div>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="form-field-2"
          >
            Hotovo:
          </label>
        </div>
        <div
          class="c5"
        >
          <div
            class="c13"
          >
            <label
              class="c14"
              for="form-field-2"
            >
              ✔
              <input
                checked=""
                class="c15"
                id="form-field-2"
                type="checkbox"
              />
            </label>
          </div>
        </div>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="c16"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="form-field-3"
          >
            Popis úkolu:
          </label>
        </div>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <textarea
              class="c17"
              id="form-field-3"
              maxlength="500"
            />
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`components/todo-detail/TodoFields > should render props 1`] = `
.c13 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-style: normal;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c10 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c11 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  padding-right: 5px;
  background: #f0f0f0;
}

.c11:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c11:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c11:disabled {
  color: #c0c0c0;
}

.c12 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c9 {
  position: relative;
  display: inline-block;
}

.c0 {
  opacity: 0.5;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  border-bottom: 1px solid #c0c0c0;
}

.c1:last-child {
  border-bottom: none;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  color: #222;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 10px;
  height: 30px;
}

.c15 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  width: 20px;
  height: 20px;
  border: 2px solid #222;
  color: #222;
  background: #fff;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
}

.c15:hover {
  background-color: #fcfcfc;
}

.c15:focus-within {
  outline: 1px dotted #222;
}

.c16 {
  position: absolute;
  top: 0;
  left: 0;
  width: 0;
  height: 0;
  opacity: 0;
  outline: none;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c7 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  color: #c22;
  border: none;
  outline: none;
}

.c7:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c7:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c7:disabled {
  color: #c0c0c0;
}

.c18 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
  min-height: 110px;
  resize: vertical;
}

.c18:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c18:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c18:disabled {
  color: #c0c0c0;
}

.c8 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  padding: 0 10px;
  color: #fff;
  background: #c22;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: block;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  gap: 10px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100px;
  min-height: 30px;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <ul
    class="c0"
  >
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="form-field-4"
          >
            Název úkolu:
          </label>
        </div>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <input
              class="c7"
              id="form-field-4"
              maxlength="80"
              type="text"
              value="Edited todo"
            />
            <div
              class="c8"
            >
              Title field error
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="form-field-5"
          >
            Složka:
          </label>
        </div>
        <div
          class="c5"
        >
          <div
            class="c9"
          >
            <button
              class="c10 c11"
              id="form-field-5"
              type="button"
            >
              <span
                class="c12"
              >
                Folder B
              </span>
              <i
                class="c13"
              >
                ˅
              </i>
            </button>
          </div>
        </div>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="form-field-6"
          >
            Hotovo:
          </label>
        </div>
        <div
          class="c5"
        >
          <div
            class="c14"
          >
            <label
              class="c15"
              for="form-field-6"
            >
              ✔
              <input
                checked=""
                class="c16"
                id="form-field-6"
                type="checkbox"
              />
            </label>
          </div>
        </div>
      </div>
    </li>
    <li
      class="c1"
    >
      <div
        class="c17"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="form-field-7"
          >
            Popis úkolu:
          </label>
        </div>
        <div
          class="c5"
        >
          <div
            class="c6"
          >
            <textarea
              class="c18"
              id="form-field-7"
              maxlength="500"
            />
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
